net/portmapper/upnp.go

```diff
@@ -130,40 +132,70 @@ func addAnyPortMapping(
 	return externalPort, err
 }
 
-// getUPnPClients gets a client for interfacing with UPnP, ignoring the underlying protocol for
+// getUPnPClient gets a client for interfacing with UPnP, ignoring the underlying protocol for
 // now.
 // Adapted from https://github.com/huin/goupnp/blob/master/GUIDE.md.
-func getUPnPClient(ctx context.Context, gw netaddr.IP) (upnpClient, error) {
+//
+// The gw is the detected gateway.
+//
+// The uPnPRootIPP is the most recently heard Location header's
+// ip:port from a UPnP discovery packet.
+func getUPnPClient(ctx context.Context, gw netaddr.IP, uPnPRootIPP netaddr.IPPort) (upnpClient, error) {
 	if controlknobs.DisableUPnP() {
 		return nil, nil
 	}
-	ctx, cancel := context.WithTimeout(ctx, 250*time.Millisecond)
-	defer cancel()
-	// Attempt to connect over the multiple available connection types concurrently,
-	// returning the fastest.
 
-	// TODO(jknodt): this url seems super brittle? maybe discovery is better but this is faster
-	u, err := url.Parse(fmt.Sprintf("http://%s:5000/rootDesc.xml", gw))
+	var port uint16
+	if uPnPRootIPP.IsZero() {
+		port = 5000
+	} else {
+		port = uPnPRootIPP.Port()
+		if uPnPRootIPP.IP() != gw {
+			return nil, fmt.Errorf("UPnP discovered root %v != gateway %v; ignoring UPnP",
+				uPnPRootIPP, gw)
+		}
+	}
+
+	rootDescURL := fmt.Sprintf("http://%s/rootDesc.xml", net.JoinHostPort(gw.String(), fmt.Sprint(port)))
```

Comment:
One other thing I noticed is that I think the caps on `rootDesc` is not capitalized the same way across all UPnP platforms, just something to note for the future

Reply:
Fixed.